### PR TITLE
Kernel IR refactoring: part 2

### PR DESCRIPTION
**Pull request by tlemo**
This iteration splits all the IR nodes into Fusion IR and Kernel IR versions. Some of the nodes are similar (ex. BinaryOp) so they exist in both IRs, although they will be more specialized and simpler as a result.

At this state there's some extra duplication reaching things like dispatch visitors. This is required to allow breaking down the refactoring into smaller pieces (in the end these changes will be removed). The alternative is a huge monolithic refactoring.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/mutator.cpp [hunk 1]
```diff
@@ -212,17 +212,7 @@ Statement* OptOutMutator::mutate(ReductionOp* rop) {
 }
 
 Statement* OptOutMutator::mutate(kir::GridReduction* gr) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/mutator.cpp:215]: I assume this means OptOutMutator is just a dispatch that can return a statement? Do we want to either do all of these at the same time or leave them as is?

**Review by csarofeen — COMMENTED** [review]
Mostly seems fine to me, would like Naoya to review too.

File: torch/csrc/jit/codegen/cuda/mutator.cpp (hunk 1 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/mutator.cpp:215]: These are just placeholders, anticipating that we're not going to use/need OptOutMutator for Kernel IR nodes. 

What do you think?

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp [hunk 2]
```diff
@@ -399,18 +473,144 @@ void IRPrinter::handle(const TernaryOp* top) {
     os << ";\n";
 }
 
-void IRPrinter::handle(const ReductionOp* rop) {
-  // Check if we've lowered yet.
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_iostream.cpp:410]: 🎉

**Review by naoyam — APPROVED** [review]
LGTM. It's very nice to have more explicit separation of the two IR stages.

**Pull request merged**